File: DB/lab_09/main.go
package main

import (
	"context"
	"database/sql"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"time"

	_ "github.com/lib/pq"
	"github.com/redis/go-redis/v9"
)

const cacheKey = "stats:top_books"

type BookStat struct {
	BookID int    `json:"book_id"`
	Title  string `json:"title"`
	Loans  int    `json:"loans"`
}

func main() {
	var (
		pgDSN           = flag.String("pg", "postgres://postgres:postgres@localhost:5432/postgres?sslmode=disable", "Postgres DSN")
		redisAdr        = flag.String("redis", "localhost:5433", "Redis host:port (у тебя 5433:6379)")
		mode            = flag.String("mode", "none", "none|insert|delete|update")
		duration        = flag.Duration("duration", 2*time.Minute, "How long to run (e.g. 2m, 30s)")
		outCSV          = flag.String("out", "results.csv", "Output CSV file")
		measureEvery    = flag.Duration("measure_every", 5*time.Second, "Measure interval")
		mutateEvery     = flag.Duration("mutate_every", 10*time.Second, "Mutation interval (insert/delete/update)")
		cacheTTL        = flag.Duration("cache_ttl", 60*time.Second, "Redis cache TTL")
		logMutations    = flag.Bool("log_mutations", true, "Log mutation ticks")
		fillCacheOnMiss = flag.Bool("fill_cache_on_miss", true, "Fill cache when miss using already computed DB result")
	)
	flag.Parse()

	ctx := context.Background()
	runCtx, cancel := context.WithTimeout(ctx, *duration)
	defer cancel()

	db, err := sql.Open("postgres", *pgDSN)
	if err != nil {
		log.Fatalf("open postgres: %v", err)
	}
	defer db.Close()
	if err := db.PingContext(runCtx); err != nil {
		log.Fatalf("ping postgres: %v", err)
	}

	rdb := redis.NewClient(&redis.Options{Addr: *redisAdr})
	defer rdb.Close()
	if err := rdb.Ping(runCtx).Err(); err != nil {
		log.Fatalf("ping redis: %v", err)
	}

	if err := seedIfEmpty(runCtx, db); err != nil {
		log.Fatalf("seed: %v", err)
	}

	f, err := os.Create(*outCSV)
	if err != nil {
		log.Fatalf("create csv: %v", err)
	}
	defer f.Close()

	w := csv.NewWriter(f)
	defer w.Flush()

	_ = w.Write([]string{"ts", "mode", "db_ms", "redis_ms", "cache_hit"})

	if *mode != "none" {
		go func() {
			t := time.NewTicker(*mutateEvery)
			defer t.Stop()
			for {
				select {
				case <-t.C:
					if *logMutations {
						log.Printf("[MUTATE] mode=%s at %s", *mode, time.Now().Format(time.RFC3339Nano))
					}
					if err := mutate(runCtx, db, rdb, *mode); err != nil {
						log.Printf("mutate(%s) error: %v", *mode, err)
					}
				case <-runCtx.Done():
					return
				}
			}
		}()
	}

	ticker := time.NewTicker(*measureEvery)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			dbStart := time.Now()
			stats, err := queryTopBooks(runCtx, db)
			dbDur := time.Since(dbStart)
			if err != nil {
				log.Printf("db query error: %v", err)
				continue
			}

			redisDur, hit, err := measureRedisGet(runCtx, rdb)
			if err != nil {
				log.Printf("redis get error: %v", err)
			}

			if !hit && *fillCacheOnMiss {
				_ = setCache(runCtx, rdb, stats, *cacheTTL)
			}

			ts := time.Now().Format(time.RFC3339Nano)
			_ = w.Write([]string{
				ts,
				*mode,
				fmt.Sprintf("%.3f", float64(dbDur.Microseconds())/1000.0),
				fmt.Sprintf("%.3f", float64(redisDur.Microseconds())/1000.0),
				fmt.Sprintf("%t", hit),
			})
			w.Flush()

			log.Printf("[%s] db=%v redis_get=%v hit=%v", *mode, dbDur, redisDur, hit)

		case <-runCtx.Done():
			log.Printf("done, wrote %s", *outCSV)
			return
		}
	}
}

func queryTopBooks(ctx context.Context, db *sql.DB) ([]BookStat, error) {
	const q = `
SELECT b.id, b.title, COUNT(*) AS loans_count
FROM tp.book_loans bl
JOIN tp.books b ON b.id = bl.book_id
GROUP BY b.id, b.title
ORDER BY loans_count DESC
LIMIT 10;`

	rows, err := db.QueryContext(ctx, q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []BookStat
	for rows.Next() {
		var s BookStat
		if err := rows.Scan(&s.BookID, &s.Title, &s.Loans); err != nil {
			return nil, err
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

func measureRedisGet(ctx context.Context, rdb *redis.Client) (time.Duration, bool, error) {
	start := time.Now()
	b, err := rdb.Get(ctx, cacheKey).Bytes()
	dur := time.Since(start)

	if err == redis.Nil {
		return dur, false, nil
	}
	if err != nil {
		return dur, false, err
	}

	var cached []BookStat
	if err := json.Unmarshal(b, &cached); err != nil {
		return dur, false, nil
	}
	return dur, true, nil
}

func setCache(ctx context.Context, rdb *redis.Client, stats []BookStat, ttl time.Duration) error {
	b, err := json.Marshal(stats)
	if err != nil {
		return err
	}
	return rdb.Set(ctx, cacheKey, b, ttl).Err()
}

func mutate(ctx context.Context, db *sql.DB, rdb *redis.Client, mode string) error {
	switch mode {
	case "insert":
		if err := insertLoan(ctx, db); err != nil {
			return err
		}
	case "delete":
		if err := deleteOneLoan(ctx, db); err != nil {
			return err
		}
	case "update":
		if err := updateOneLoan(ctx, db); err != nil {
			return err
		}
	default:
		return nil
	}

	_ = rdb.Del(ctx, cacheKey).Err()
	return nil
}

func nextID(ctx context.Context, db *sql.DB, table string) (int, error) {
	q := fmt.Sprintf(`SELECT COALESCE(MAX(id), 0) + 1 FROM %s;`, table)
	var id int
	if err := db.QueryRowContext(ctx, q).Scan(&id); err != nil {
		return 0, err
	}
	return id, nil
}

func pickRandomIDOffset(ctx context.Context, db *sql.DB, table string) (int, error) {
	q := fmt.Sprintf(`
WITH c AS (SELECT COUNT(*)::int AS n FROM %s)
SELECT id
FROM %s
OFFSET (SELECT CASE WHEN n=0 THEN 0 ELSE floor(random()*n)::int END FROM c)
LIMIT 1;`, table, table)

	var id int
	if err := db.QueryRowContext(ctx, q).Scan(&id); err != nil {
		return 0, err
	}
	return id, nil
}

func insertLoan(ctx context.Context, db *sql.DB) error {
	loanID, err := nextID(ctx, db, "tp.book_loans")
	if err != nil {
		return err
	}
	bookID, err := pickRandomIDOffset(ctx, db, "tp.books")
	if err != nil {
		return err
	}
	authorID, err := pickRandomIDOffset(ctx, db, "tp.authors")
	if err != nil {
		return err
	}
	readerID, err := pickRandomIDOffset(ctx, db, "tp.readers")
	if err != nil {
		return err
	}

	loanDate := time.Now().Truncate(24 * time.Hour)
	dueDate := loanDate.AddDate(0, 0, 14)

	_, err = db.ExecContext(ctx, `
INSERT INTO tp.book_loans(id, book_id, author_id, reader_id, loan_date, due_date, return_date, status)
VALUES($1,$2,$3,$4,$5,$6,NULL,'Выдана');`,
		loanID, bookID, authorID, readerID, loanDate, dueDate,
	)
	return err
}

func deleteOneLoan(ctx context.Context, db *sql.DB) error {
	var id int
	err := db.QueryRowContext(ctx, `SELECT id FROM tp.book_loans ORDER BY id DESC LIMIT 1;`).Scan(&id)
	if err == sql.ErrNoRows {
		return nil
	}
	if err != nil {
		return err
	}
	_, err = db.ExecContext(ctx, `DELETE FROM tp.book_loans WHERE id=$1;`, id)
	return err
}

func updateOneLoan(ctx context.Context, db *sql.DB) error {
	var id int
	err := db.QueryRowContext(ctx, `
SELECT id
FROM tp.book_loans
WHERE status = 'Выдана'
ORDER BY id DESC
LIMIT 1;`).Scan(&id)

	if err == sql.ErrNoRows {
		err = db.QueryRowContext(ctx, `SELECT id FROM tp.book_loans ORDER BY id DESC LIMIT 1;`).Scan(&id)
		if err == sql.ErrNoRows {
			return nil
		}
	}
	if err != nil {
		return err
	}

	returnDate := time.Now().Truncate(24 * time.Hour)
	_, err = db.ExecContext(ctx, `
UPDATE tp.book_loans
SET status='Возвращена', return_date=$2
WHERE id=$1;`, id, returnDate)
	return err
}

func seedIfEmpty(ctx context.Context, db *sql.DB) error {
	var cnt int
	if err := db.QueryRowContext(ctx, `SELECT COUNT(*) FROM tp.genres;`).Scan(&cnt); err != nil {
		return err
	}
	if cnt > 0 {
		return nil
	}

	log.Printf("tables look empty -> seeding minimal data...")

	_, err := db.ExecContext(ctx, `
INSERT INTO tp.genres(id,name) VALUES
(1,'Фантастика'),(2,'Детектив'),(3,'Научпоп');`)
	if err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, `
INSERT INTO tp.authors(id,first_name,last_name,birth_date,country,biography) VALUES
(1,'Айзек','Азимов','1920-01-02','USA',''),
(2,'Артур','Конан Дойл','1859-05-22','UK',''),
(3,'Стивен','Хокинг','1942-01-08','UK','');`)
	if err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, `
INSERT INTO tp.books(id,title,genre_id,publication_year,publisher,page_count) VALUES
(1,'Основание',1,1951,'Gnome Press',255),
(2,'Шерлок Холмс',2,1892,'George Newnes',307),
(3,'Краткая история времени',3,1988,'Bantam Dell',256);`)
	if err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, `
INSERT INTO tp.readers(id,surname,name,middle_name,address,sex,birthday,email,phone,registration_date) VALUES
(1,'Иванов','Иван','Иванович','Москва','М','2000-01-01','ivanov@example.com','+70000000001',CURRENT_DATE),
(2,'Петрова','Анна','Сергеевна','СПб','Ж','1999-02-02','petrova@example.com','+70000000002',CURRENT_DATE);`)
	if err != nil {
		return err
	}

	rand.Seed(time.Now().UnixNano())
	for i := 1; i <= 200; i++ {
		bookID := 1 + rand.Intn(3)
		authorID := 1 + rand.Intn(3)
		readerID := 1 + rand.Intn(2)

		loanDate := time.Now().AddDate(0, 0, -rand.Intn(60)).Truncate(24 * time.Hour)
		dueDate := loanDate.AddDate(0, 0, 14)

		_, err := db.ExecContext(ctx, `
INSERT INTO tp.book_loans(id, book_id, author_id, reader_id, loan_date, due_date, return_date, status)
VALUES($1,$2,$3,$4,$5,$6,NULL,'Выдана');`,
			i, bookID, authorID, readerID, loanDate, dueDate,
		)
		if err != nil {
			return err
		}
	}

	log.Printf("seed done")
	return nil
}

File: DB/lab_09/main.py
import os
import json
import time
import csv
from random import randint, choice

import psycopg2
import redis
import matplotlib.pyplot as plt


# ----------------- CONFIG -----------------
PG_DSN = os.getenv(
    "PG_DSN",
    "dbname=postgres user=postgres password=postgres host=127.0.0.1 port=5432"
)
REDIS_HOST = os.getenv("REDIS_HOST", "localhost")
REDIS_PORT = int(os.getenv("REDIS_PORT", "5433"))  # у тебя 5433:6379
CACHE_KEY = "stats:top_books"
CACHE_TTL_SEC = int(os.getenv("CACHE_TTL_SEC", "60"))

QUERY_INTERVAL_SEC = 5
MUTATE_INTERVAL_SEC = 10


# ----------------- DB / REDIS -----------------
def connection():
    try:
        con = psycopg2.connect(PG_DSN)
        con.autocommit = True
    except Exception as e:
        print("Ошибка при подключении к Postgres:", e)
        raise
    print("Postgres подключен")
    return con


def redis_client():
    r = redis.Redis(host=REDIS_HOST, port=REDIS_PORT, db=0)
    # ping для быстрой проверки
    r.ping()
    return r


# ----------------- STAT QUERY (Library theme) -----------------
STAT_SQL = """
SELECT b.id, b.title, COUNT(*) AS loans_count
FROM tp.book_loans bl
JOIN tp.books b ON b.id = bl.book_id
GROUP BY b.id, b.title
ORDER BY loans_count DESC
LIMIT 10;
"""


def query_top_books_db(cur):
    """Возвращает список списков: [[book_id, title, loans_count], ...]"""
    cur.execute(STAT_SQL)
    rows = cur.fetchall()
    return [[int(r[0]), str(r[1]), int(r[2])] for r in rows]


def get_top_books_cached(cur, r):
    """
    Возвращает (result, hit, db_ms, redis_ms)
    Требование "мерить время самих запросов":
      - db_ms: время SQL (execute+fetchall)
      - redis_ms: время Redis GET (+json.loads при hit)
    При miss мы НЕ делаем второй DB-запрос: используем уже измеренный db_result.
    """
    # 1) мерим DB запрос (как эталон)
    t1 = time.perf_counter()
    db_result = query_top_books_db(cur)
    t2 = time.perf_counter()
    db_ms = (t2 - t1) * 1000.0

    # 2) мерим Redis GET
    t3 = time.perf_counter()
    blob = r.get(CACHE_KEY)
    if blob is not None:
        # hit: декодируем
        result = json.loads(blob.decode("utf-8"))
        t4 = time.perf_counter()
        redis_ms = (t4 - t3) * 1000.0
        return result, True, db_ms, redis_ms

    # miss: redis GET тоже измерен, но данных нет
    t4 = time.perf_counter()
    redis_ms = (t4 - t3) * 1000.0

    # заполняем кэш уже готовым db_result (не портим db_ms)
    r.setex(CACHE_KEY, CACHE_TTL_SEC, json.dumps(db_result, ensure_ascii=False).encode("utf-8"))
    return db_result, False, db_ms, redis_ms


# ----------------- MUTATIONS (every 10s) -----------------
def get_max_id(cur, table):
    cur.execute(f"SELECT COALESCE(MAX(id), 0) FROM {table};")
    return int(cur.fetchone()[0])


def pick_existing_id(cur, table):
    """
    Без ORDER BY RANDOM() (он шумит).
    Берём диапазон id и пытаемся найти существующий.
    """
    max_id = get_max_id(cur, table)
    if max_id <= 0:
        return None
    for _ in range(20):
        cand = randint(1, max_id)
        cur.execute(f"SELECT id FROM {table} WHERE id=%s;", (cand,))
        row = cur.fetchone()
        if row:
            return int(row[0])
    # fallback: последняя строка
    cur.execute(f"SELECT id FROM {table} ORDER BY id DESC LIMIT 1;")
    row = cur.fetchone()
    return int(row[0]) if row else None


def mutate_insert(cur, r):
    # создаём новую выдачу
    new_id = get_max_id(cur, "tp.book_loans") + 1

    book_id = pick_existing_id(cur, "tp.books")
    author_id = pick_existing_id(cur, "tp.authors")
    reader_id = pick_existing_id(cur, "tp.readers")
    if None in (book_id, author_id, reader_id):
        return

    # loan_date сегодня, due_date +14
    cur.execute("""
        INSERT INTO tp.book_loans(id, book_id, author_id, reader_id, loan_date, due_date, return_date, status)
        VALUES(%s,%s,%s,%s,CURRENT_DATE, CURRENT_DATE + INTERVAL '14 days', NULL, 'Выдана');
    """, (new_id, book_id, author_id, reader_id))

    # инвалидируем кэш статистики
    r.delete(CACHE_KEY)


def mutate_delete(cur, r):
    # удаляем последнюю выдачу (стабильно и без RANDOM)
    cur.execute("SELECT id FROM tp.book_loans ORDER BY id DESC LIMIT 1;")
    row = cur.fetchone()
    if not row:
        return
    loan_id = int(row[0])
    cur.execute("DELETE FROM tp.book_loans WHERE id=%s;", (loan_id,))
    r.delete(CACHE_KEY)


def mutate_update(cur, r):
    # закрываем (возвращаем) последнюю "Выдана"
    cur.execute("""
        SELECT id FROM tp.book_loans
        WHERE status='Выдана'
        ORDER BY id DESC
        LIMIT 1;
    """)
    row = cur.fetchone()
    if not row:
        return
    loan_id = int(row[0])

    cur.execute("""
        UPDATE tp.book_loans
        SET status='Возвращена', return_date=CURRENT_DATE
        WHERE id=%s;
    """, (loan_id,))
    r.delete(CACHE_KEY)


def do_mutation(cur, r, mode):
    if mode == "insert":
        mutate_insert(cur, r)
    elif mode == "delete":
        mutate_delete(cur, r)
    elif mode == "update":
        mutate_update(cur, r)
    # mode none -> no-op


# ----------------- EXPERIMENT LOOP -----------------
def run_experiment(mode: str, duration_sec: int, out_csv: str):
    """
    mode: none|insert|delete|update
    duration_sec: сколько секунд крутить эксперимент
    Каждые 5 сек: замер
    Каждые 10 сек: мутация (если mode != none)
    """
    con = connection()
    cur = con.cursor()
    r = redis_client()

    # лучше начать с чистого ключа
    r.delete(CACHE_KEY)

    rows = []
    start = time.time()
    next_query = start
    next_mutate = start

    while True:
        now = time.time()
        if now - start >= duration_sec:
            break

        # 10 секунд: мутация
        if mode != "none" and now >= next_mutate:
            do_mutation(cur, r, mode)
            next_mutate += MUTATE_INTERVAL_SEC

        # 5 секунд: замер
        if now >= next_query:
            result, hit, db_ms, redis_ms = get_top_books_cached(cur, r)
            # result можно не печатать, но оставлю при желании:
            # print("TOP:", result)

            rows.append({
                "ts": time.strftime("%Y-%m-%dT%H:%M:%S", time.localtime(now)),
                "mode": mode,
                "db_ms": round(db_ms, 3),
                "redis_ms": round(redis_ms, 3),
                "cache_hit": 1 if hit else 0
            })

            print(f"[{mode}] db={db_ms:.3f}ms redis={redis_ms:.3f}ms hit={hit}")
            next_query += QUERY_INTERVAL_SEC

        time.sleep(0.02)

    # write CSV
    with open(out_csv, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=["ts", "mode", "db_ms", "redis_ms", "cache_hit"])
        w.writeheader()
        w.writerows(rows)

    cur.close()
    con.close()
    r.close()
    print("Готово:", out_csv)
    return out_csv


# ----------------- PLOTS -----------------
def plot_lines(csv_file: str, title: str):
    xs, dbs, reds = [], [], []
    with open(csv_file, "r", encoding="utf-8") as f:
        rd = csv.DictReader(f)
        for i, row in enumerate(rd):
            xs.append(i)
            dbs.append(float(row["db_ms"]))
            reds.append(float(row["redis_ms"]))

    plt.figure()
    plt.plot(xs, dbs, marker="o", label="Postgres (SQL time)")
    plt.plot(xs, reds, marker="s", label="Redis (GET+decode time)")
    plt.title(title)
    plt.xlabel("Measurement index")
    plt.ylabel("ms")
    plt.grid(True)
    plt.legend()
    plt.show()


def plot_bars_avg(csv_file: str, title: str):
    dbs, reds = [], []
    with open(csv_file, "r", encoding="utf-8") as f:
        rd = csv.DictReader(f)
        for row in rd:
            dbs.append(float(row["db_ms"]))
            reds.append(float(row["redis_ms"]))

    avg_db = sum(dbs) / len(dbs) if dbs else 0
    avg_redis = sum(reds) / len(reds) if reds else 0

    plt.figure()
    plt.bar(["DB", "Redis"], [avg_db, avg_redis])
    plt.title(title)
    plt.ylabel("Average ms")
    plt.grid(True, axis="y")
    plt.show()


# ----------------- MAIN -----------------
if __name__ == "__main__":
    print(
        "Режимы:\n"
        " 1) none   - без изменения данных\n"
        " 2) insert - добавление строки каждые 10 секунд\n"
        " 3) delete - удаление строки каждые 10 секунд\n"
        " 4) update - изменение строки каждые 10 секунд\n"
    )

    mode = input("Введи режим (none/insert/delete/update): ").strip()
    duration = int(input("Длительность эксперимента (сек), напр. 180: ").strip())

    out = f"{mode}.csv"
    run_experiment(mode=mode, duration_sec=duration, out_csv=out)

    plot_lines(out, f"Mode: {mode} (lines)")
    plot_bars_avg(out, f"Mode: {mode} (average)")
